grudge/trace_pair.py

```diff
@@ -394,33 +389,35 @@ def connected_parts(
         dcoll._volume_discrs[volume_dd.domain_tag.tag].mesh)
 
 
-def _sym_tag_to_num_tag(comm_tag: Hashable | None) -> int | None:
+def _sym_tag_to_num_tag(comm_tag: Hashable | None, base_tag: int) -> int | None:
     if comm_tag is None:
-        return comm_tag
+        return base_tag
 
     if isinstance(comm_tag, int):
-        return comm_tag
+        return comm_tag + base_tag
 
     # FIXME: This isn't guaranteed to be correct.
     # See here for discussion:
     # - https://github.com/illinois-ceesd/mirgecom/issues/617#issuecomment-1057082716  # noqa
     # - https://github.com/inducer/grudge/pull/222
+    # Since only 1 communication can be pending for a given tag at a time,
+    # this does not matter currently. See https://github.com/inducer/grudge/issues/223
 
     from mpi4py import MPI
     tag_ub = MPI.COMM_WORLD.Get_attr(MPI.TAG_UB)
     assert tag_ub is not None
-    key_builder = _TagKeyBuilder()
+    key_builder = KeyBuilder()
     digest = key_builder(comm_tag)
 
     num_tag = sum(ord(ch) << i for i, ch in enumerate(digest)) % tag_ub
 
     warn("Encountered unknown symbolic tag "
```

Comment:
Perhaps we should downgrade this to logger.info?

Reply:
Nope. It's a correctness issue that would be straightforward to avoid.